sstables/generation_type.hh

```diff
@@ -49,6 +49,26 @@ inline generation_type new_generation(std::optional<generation_type> prev = std:
     }
 }
 
+class generation_factory {
+    std::optional<generation_type> _next;
+    std::optional<generation_type> _prev;
+public:
+    generation_factory(generation_type gen = new_generation()) noexcept : _next(gen) {}
+
+    generation_factory(const generation_factory&) = delete;
+    generation_factory(generation_factory&&) = delete;
+
+    generation_type prev() const noexcept {
+        return *_prev;
+    }
+
+    generation_type operator()() noexcept {
+        _prev = _next;
+        _next = new_generation(_next);
+        return *_prev;
+    }
+};
```

Comment:
just a head-up, this is likely to conflict with the change in https://github.com/scylladb/scylladb/pull/13073/commits/1aa802f8c5a5e9175b5a2452691003b8644eea0f .

Reply:
I suggest this PR should be merged first as it's likely to bit-rot otherwise, and it provides some basic functionality that can be used by later changes to the product code.

Reply:
absolutely. i will take another look at this PR. my changes are based on this changeset. the early we have this PR landed the better.

Reply:
Again I don't see why we need a generation_factory for something so simple. Can't it just be part of the code that gets a new generation?

Reply:
Unfortunately we have several places that generate new generations in the code, especially in tests.
I wanted whoever needs a new generation to get one in abstract way, not caring if the factory dispatches integer generations or a uuid ones.
With such a class we can also specialize it to add smp::count to the internal counter for table shards (until uuid generations)